File: ios/Halogen/Core/NavConfig.swift
import Foundation
import Observation

/// Built-in navigation destinations — mirrors the web's `BuiltinNav`
/// (crates/ui-config nav.rs), same snake_case tokens so a future config sync
/// could share them.
enum BuiltinNav: String, Codable, CaseIterable, Identifiable {
    case queue
    case latest
    case podcasts
    case playlists
    case downloads
    case discover
    case history
    case settings
    case polling
    case deviceLogs = "device_logs"
    case serverLogs = "server_logs"

    var id: String { rawValue }

    var label: String {
        switch self {
        case .queue: return "Queue"
        case .latest: return "Latest"
        case .podcasts: return "Podcasts"
        case .playlists: return "Playlists"
        case .downloads: return "Downloads"
        case .discover: return "Discover"
        case .history: return "History"
        case .settings: return "Settings"
        case .polling: return "Polling"
        case .deviceLogs: return "Device Logs"
        case .serverLogs: return "Server Logs"
        }
    }

    /// Admin-only destinations (web: `NavItem.admin_only` — Polling and
    /// Server Logs are dropped from every nav surface for non-admins).
    var adminOnly: Bool {
        switch self {
        case .polling, .serverLogs: return true
        default: return false
        }
    }

    var systemImage: String {
        switch self {
        case .queue: return "list.bullet"
        case .latest: return "clock"
        case .podcasts: return "square.grid.2x2"
        case .playlists: return "music.note.list"
        case .downloads: return "arrow.down.circle"
        case .discover: return "magnifyingglass"
        case .history: return "clock.arrow.circlepath"
        case .settings: return "gear"
        case .polling: return "arrow.triangle.2.circlepath"
        case .deviceLogs: return "doc.text"
        case .serverLogs: return "server.rack"
        }
    }
}

/// Nav order/visibility — the web's `NavConfig` shape (order + hidden +
/// pinned playlist ids), persisted per account. The dock renders the first
/// `dockSlots` visible items plus the always-present More tab; the More menu
/// renders everything visible.
struct NavConfig: Codable, Equatable {
    var order: [BuiltinNav]
    var hidden: [BuiltinNav]
    /// Playlist ids pinned as nav links (parity field; pin UI not built yet).
    var pinnedPlaylists: [Int32]

    static let `default` = NavConfig(
        order: [
            .queue, .latest, .podcasts, .playlists, .downloads,
            .discover, .history, .settings, .polling, .deviceLogs, .serverLogs,
        ],
        hidden: [.polling, .deviceLogs, .serverLogs],
        pinnedPlaylists: []
    )

    /// Self-heal persisted orders that predate newly added builtins.
    mutating func normalize() {
        for key in Self.default.order where !order.contains(key) {
            order.append(key)
        }
        // Settings can never be hidden (same hard rule as the web).
        hidden.removeAll { $0 == .settings }
    }

    var visible: [BuiltinNav] {
        order.filter { !hidden.contains($0) }
    }
}

/// Reactive holder + persistence for the account's NavConfig.
@MainActor
@Observable
final class NavModel {
    /// iPhone shows at most 5 tab slots and the More tab always takes one, so
    /// 4 destinations ride the dock (the web dock fits 5 + More; platform
    /// divergence documented in docs/internal/IOS_NATIVE_SWIFT.md).
    static let dockSlots = 4
    private static let key = "nav-config"

    private(set) var config: NavConfig = .default
    private let store: LocalStore?

    init(store: LocalStore?) {
        self.store = store
    }

    func load() async {
        if let store, var saved = await store.load(NavConfig.self, key: Self.key) {
            saved.normalize()
            config = saved
        }
    }

    func update(_ new: NavConfig) {
        var normalized = new
        normalized.normalize()
        config = normalized
        let snapshot = normalized
        Task { [store] in await store?.save(snapshot, key: Self.key) }
    }

    /// Every destination this user may see, in configured order — the web's
    /// `nav_items`: hidden dropped, admin-only dropped for non-admins.
    func visibleItems(isAdmin: Bool) -> [BuiltinNav] {
        config.visible.filter { isAdmin || !$0.adminOnly }
    }

    /// The tab bar: first N visible destinations (More is appended by the UI).
    func dockItems(isAdmin: Bool) -> [BuiltinNav] {
        Array(visibleItems(isAdmin: isAdmin).prefix(Self.dockSlots))
    }
}

File: android/app/src/main/kotlin/org/fgsec/halogen/core/NavConfig.kt
package org.fgsec.halogen.core

import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.MainScope
import kotlinx.coroutines.launch
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import org.fgsec.halogen.storage.LocalStore

/// Built-in navigation destinations — mirrors the web's `BuiltinNav`
/// (webui/config nav.rs), same snake_case tokens so a future config sync
/// could share them.
@Serializable
enum class BuiltinNav(val token: String) {
    @SerialName("queue") Queue("queue"),
    @SerialName("latest") Latest("latest"),
    @SerialName("podcasts") Podcasts("podcasts"),
    @SerialName("playlists") Playlists("playlists"),
    @SerialName("downloads") Downloads("downloads"),
    @SerialName("discover") Discover("discover"),
    @SerialName("history") History("history"),
    @SerialName("settings") Settings("settings"),
    @SerialName("polling") Polling("polling"),
    @SerialName("device_logs") DeviceLogs("device_logs"),
    @SerialName("server_logs") ServerLogs("server_logs");

    val label: String
        get() = when (this) {
            Queue -> "Queue"
            Latest -> "Latest"
            Podcasts -> "Podcasts"
            Playlists -> "Playlists"
            Downloads -> "Downloads"
            Discover -> "Discover"
            History -> "History"
            Settings -> "Settings"
            Polling -> "Polling"
            DeviceLogs -> "Device Logs"
            ServerLogs -> "Server Logs"
        }

    /// Admin-only destinations (web: `NavItem.admin_only` — Polling and
    /// Server Logs are dropped from every nav surface for non-admins).
    val adminOnly: Boolean
        get() = this == Polling || this == ServerLogs

    /// SF Symbol name — components/Icons.kt maps it to a Material icon.
    val systemImage: String
        get() = when (this) {
            Queue -> "list.bullet"
            Latest -> "clock"
            Podcasts -> "square.grid.2x2"
            Playlists -> "music.note.list"
            Downloads -> "arrow.down.circle"
            Discover -> "magnifyingglass"
            History -> "clock.arrow.circlepath"
            Settings -> "gear"
            Polling -> "arrow.triangle.2.circlepath"
            DeviceLogs -> "doc.text"
            ServerLogs -> "server.rack"
        }
}

/// Nav order/visibility — the web's `NavConfig` shape (order + hidden +
/// pinned playlist ids), persisted per account. The dock renders the first
/// `NavModel.DOCK_SLOTS` visible items plus the always-present More tab; the
/// More menu renders everything visible.
@Serializable
data class NavConfig(
    val order: List<BuiltinNav>,
    val hidden: List<BuiltinNav>,
    /// Playlist ids pinned as nav links (parity field; pin UI not built yet).
    val pinnedPlaylists: List<Int>,
) {
    /// Self-heal persisted orders that predate newly added builtins.
    /// Settings can never be hidden (same hard rule as the web).
    fun normalized(): NavConfig = copy(
        order = order + default.order.filter { it !in order },
        hidden = hidden.filter { it != BuiltinNav.Settings },
    )

    val visible: List<BuiltinNav>
        get() = order.filter { it !in hidden }

    companion object {
        val default = NavConfig(
            order = listOf(
                BuiltinNav.Queue, BuiltinNav.Latest, BuiltinNav.Podcasts,
                BuiltinNav.Playlists, BuiltinNav.Downloads, BuiltinNav.Discover,
                BuiltinNav.History, BuiltinNav.Settings, BuiltinNav.Polling,
                BuiltinNav.DeviceLogs, BuiltinNav.ServerLogs,
            ),
            hidden = listOf(BuiltinNav.Polling, BuiltinNav.DeviceLogs, BuiltinNav.ServerLogs),
            pinnedPlaylists = emptyList(),
        )
    }
}

/// Reactive holder + persistence for the account's NavConfig.
class NavModel(
    private val store: LocalStore?,
    private val scope: CoroutineScope = MainScope(),
) {
    var config: NavConfig by mutableStateOf(NavConfig.default)
        private set

    suspend fun load() {
        val saved = store?.load<NavConfig>(KEY) ?: return
        config = saved.normalized()
    }

    fun update(new: NavConfig) {
        val normalized = new.normalized()
        config = normalized
        scope.launch { store?.save(normalized, KEY) }
    }

    /// Every destination this user may see, in configured order — the web's
    /// `nav_items`: hidden dropped, admin-only dropped for non-admins.
    fun visibleItems(isAdmin: Boolean): List<BuiltinNav> =
        config.visible.filter { isAdmin || !it.adminOnly }

    /// The tab bar: first N visible destinations (More is appended by the UI).
    fun dockItems(isAdmin: Boolean): List<BuiltinNav> =
        visibleItems(isAdmin).take(DOCK_SLOTS)

    companion object {
        /// 4 destinations + More, as on iOS. Six slots fit physically but
        /// not legibly — the labels ellipsized.
        const val DOCK_SLOTS = 4
        private const val KEY = "nav-config"
    }
}
